### PR TITLE
[CI] Make inductor-perf-test-nightly produce data for dashboard

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95685

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

**Comment by desertfire** [comment]
xref https://github.com/pytorch/pytorch/pull/95675

File: .ci/pytorch/common_utils.sh [hunk 1]
```diff
@@ -153,6 +153,14 @@ function install_filelock() {
   pip_install filelock
 }
 
+function install_matplotlib() {
+  pip_install matplotlib
```
Comment (huydhn) [inline .ci/pytorch/common_utils.sh:157]: If you are looking to reduce the chance of flakiness coming from pip installation, I think it's better to add these two dependencies into https://github.com/pytorch/pytorch/blob/master/.ci/docker/requirements-ci.txt instead.  They will become part of the Docker image used for this the perf job.  Then `install_*` can be removed from the CI.

Comment (desertfire) [reply .ci/pytorch/common_utils.sh:157]: Good point. Is there a specific image config I should add to? These libs are not needed for every single test, so I think we should avoid exploding the common image?

Comment (huydhn) [reply .ci/pytorch/common_utils.sh:157]: You can tweak this if you want to.  The `requirements-ci.txt` file is for all images and adding 1 or 2 dependencies there is fine.  

FYI, The one you need is https://github.com/pytorch/pytorch/blob/master/.github/workflows/inductor-perf-test-nightly.yml#L24 (`pytorch-linux-bionic-cuda11.7-cudnn8-py3-gcc7`).  And you can tweak it logic in https://github.com/pytorch/pytorch/blob/master/.ci/docker/build.sh#L104-L116, i.e. having a param like `WITH_DASHBOARD_DEP=yes` then tweak the shared Dockerfile accordingly https://github.com/pytorch/pytorch/blob/master/.ci/docker/ubuntu-cuda/Dockerfile#L124.  But it looks rather complicated IMO.  A slightly different way is to create a brand new Docker image, but  the same principal applies.

Comment (desertfire) [reply .ci/pytorch/common_utils.sh:157]: OK, sounds like not worth the effort for small libs. But on a related note, the other request for having an image with preinstalled benchmarks probably needs to have its own image.

File: .ci/pytorch/test.sh [hunk 2]
```diff
@@ -364,8 +362,11 @@ test_inductor_benchmark_perf() {
         --expected benchmarks/dynamo/expected_ci_perf_inductor_torchbench.csv
     done
   else
-    python benchmarks/dynamo/$1.py --ci --training --performance --disable-cudagraphs\
-      --device cuda --inductor --amp $PARTITION_FLAGS  --output "$TEST_REPORTS_DIR"/inductor_training_$1.csv
+    # MKL_THREADING_LAYER=GNU to mitigate https://github.com/pytorch/pytorch/issues/37377
+    MKL_THREADING_LAYER=GNU python benchmarks/dynamo/runner.py --suites=$1 --training --dtypes=amp \
+      --base-sha="$BASE_SHA" --output-dir="$TEST_REPORTS_DIR" --dashboard-archive-path="$TEST_ARCHIVE_DIR"
+    MKL_THREADING_LAYER=GNU python benchmarks/dynamo/runner.py --suites=$1 --training --dtypes=float32 \
```
Comment (weiwangmeta) [inline .ci/pytorch/test.sh:368]: Does this make the mkl issue go away?

**Review by weiwangmeta — COMMENTED** [review]
Overall, did the job's total execution increase? They used to be able to finish in ~one hour each (total 4-5hours). Is now the time 6+ hours?

**Comment by desertfire** [comment]
> Overall, did the job's total execution increase? They used to be able to finish in ~one hour each (total 4-5hours). Is now the time 6+ hours?

Yes, I had to add `timeout-minutes: 480`. The default 240 causes timeout.

File: .ci/pytorch/test.sh (hunk 2 above)
Comment (desertfire) [reply .ci/pytorch/test.sh:368]: Yes, it does.

Comment (desertfire) [reply .ci/pytorch/test.sh:368]: But we need a proper fix as discussed in https://github.com/pytorch/pytorch/issues/37377 (which suggests the problem was already solved, but I am not sure why it shows up again).

**Comment by weiwangmeta** [comment]
Could you please do a rebase?  The inductor-A100-perf seems to have been triggering duplicate runs. After rebase, each push should just trigger one workflow run.

File: .ci/pytorch/test.sh (hunk 2 above)
Comment (weiwangmeta) [reply .ci/pytorch/test.sh:368]: Agreed.

**Comment by weiwangmeta** [comment]
We need to compare and note down the end to end time increase. With workflow duration of ~4-5hours, the runners are barely handling the 3 times a day scheduled jobs and perf smoke tests per commit. Each use of inductor/a100-perf-test label would trigger additional jobs. 

I observe this workflow https://github.com/pytorch/pytorch/actions/runs/4297691511/usage is now taking total 11+ hrs (at least 3 hour wait assuming 3 runners are used to serve this traffic). The infra might not be able to handle this increase and we need to draft a plan to be able to enable this.

File: .ci/pytorch/common_utils.sh (hunk 1 above)
Comment (ezyang) [reply .ci/pytorch/common_utils.sh:157]: IMO, it is worth doing even for small libs. The big benefit is that a version upgrade from upstream doesn't silently break our CI.

File: .github/workflows/inductor-perf-test-nightly.yml [hunk 3]
```diff
@@ -23,8 +23,7 @@ jobs:
       test-matrix: |
         { include: [
           { config: "inductor_huggingface_perf", shard: 1, num_shards: 1, runner: "linux.gcp.a100" },
-          { config: "inductor_timm_perf", shard: 1, num_shards: 2, runner: "linux.gcp.a100" },
-          { config: "inductor_timm_perf", shard: 2, num_shards: 2, runner: "linux.gcp.a100" },
+          { config: "inductor_timm_perf", shard: 1, num_shards: 1, runner: "linux.gcp.a100" },
```
Comment (ezyang) [inline .github/workflows/inductor-perf-test-nightly.yml:26]: why unshard?

**Review by ezyang — APPROVED** [review]
This seems fine, I'll approve to unblock, but the loss of sharding stings

**Comment by desertfire** [comment]
> This seems fine, I'll approve to unblock, but the loss of sharding stings

We need to add sharding to runner.py for that purpose, but the question boils down to what we aim to use this inductor-perf-test-nightly for. If it is just for nightly, then we are fine here. But if we also want this for testing PR perf difference, we need an even finer sharding given the number of configs we are testing in a single runner.py run.

I propose to create another workflow for PR perf testing. We should only test one config (probably just amp training) for faster turnaround time.

**Comment by ezyang** [comment]
For PR perf testing, I agree defaulting to AMP only is good, but I think people in some situations will want to test other cases too

**Comment by weiwangmeta** [comment]
Hi Bin, is the end to end time increase to 18+ hours expected? That seems about a 3x slowdown. Did the job start to run for all 4 backends?

**Comment by desertfire** [comment]
> Hi Bin, is the end to end time increase to 18+ hours expected? That seems about a 3x slowdown. Did the job start to run for all 4 backends?

Yes, it is expected. It includes all the combinations of (accuracy, perf) x (fp32, AMP) x (eager, aot_eager, inductor_cudagraph, inductor_nocudagraph). I think as long as the execution is within 24 hours, we are ok, because we are testing "nightly".

**Comment by weiwangmeta** [comment]
I see, we are turning this into full dashboard run.

**Review by huydhn — APPROVED** [review]
Pre-stamp approval while waiting for additional A100 runners.  I think we should also name the runner differently here, i.e. calling `linux.gcp.a100.perf` or something instead of `linux.gcp.a100` per the discussion today

**Comment by weiwangmeta** [comment]
I'm working on changing it to linux.gcp.a100.large, ETA by end of today
